rfc/rfc-0000-common-blueprint.md

```diff
@@ -0,0 +1,220 @@
+* `ClusterBlueprints` will encapsulate everything the current templates allow, while adding `spec.resources` to enable defining multiple resources (i.e. snippets).
+* A `ClusterBlueprint` can take one of `spec.resources`, `spec.template`, and `spec.ytt`. It will allow inlining of templates in `spec.resources`.
+* A `ClusterBlueprint` must specify a `blueprintTypeRef` which defines the outputs required from the blueprint. The `ClusterBlueprint`'s `spec.outputs` must match those outputs.
+* A `ClusterBlueprintType` defines only `spec.outputs`, which is an array of output names. By leaving the `name` key in the array, we leave open the possibility to later add `type` to each output, though this RFC currently proposes only supporting string outputs.
+* A `ClusterBlueprintSelector` defines a `spec.blueprintRef`, `spec.selector` (per the v1alpha2 syntax), and `spec.ownerType` to define the type of owner to match against. This type takes an `apiVersion` and `kind` to leave open the possibility to later support other owner types, though this RFC proposes only supporting the existing workload and deliverable types.
+
+Example with resources, inlining, two outputs, and workload selector:
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterBlueprint
+metadata:
+  name: my-blueprint
+spec:
+  blueprintTypeRef:
+    name: my-baz-fizz-type
+  
+  # mutually exclusive - resources, template, ytt 
+  resources:
+    - name: first-resource
+      inputs:
+        - resource: zero-resource
+          output: zoo
+          name: zoo-resource-name
+      inline:
+        # mutually exclusive - template, ytt
+        template: {}
+        ytt: {}
+
+        outputs:
+          - name: foo
+            valuePath: .status.url
+
+    - name: second-resource
+      # mutually exclusive - blueprintName, options
+      options:
+        - name: another-blueprint
+        - name: diff-blueprint
+      inputs:
+        - resource: first-resource
+          output: foo
+          name: a-new-resource-name
+        - resource: zero-resource
+          output: zoo
+          name: a-new-resource-name
+
+    - name: third-resource
+      # mutually exclusive - blueprintName, options
+      blueprintName: another-another-blueprint
+      inputs:
+        - resource: first-resource
+          output: foo
+          name: a-new-resource-name
+        - resource: second-resource
+          output: bar
+          name: another-new-name
+  
+  outputs:
+    - name: baz
+      valuePath: resources[first-resource].outputs[foo]
+    - name: fizz
+      valuePath: resources[second-resource].outputs[bar]
+
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterBlueprintType
+metadata:
+  name: my-baz-fizz-type
+spec:
+  outputs:
+    - name: baz
+    - name: fizz
+  
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterBlueprintSelector
+metadata:
+  name: my-selector
+spec:
+  blueprintRef:
+    name: my-blueprint
+  ownerType:
+    apiVersion: carto.run/v1alpha1
+    kind: Workload
```

Comment:
Do we need another field that defines the param name for template interpolation? Or maybe `$(workload)$` and `$(deliverable)$` get replaced with something generic (like `$(descriptor)$`)?

I could see both approaches working, but the former would allow template re-use across descriptors with common fields.

Reply:
I think another field could make sense, otherwise `owner` seems sensible.

Reply:
@scothis what do you mean? ownerTypes and outputs would need to be available in the context for stamping.

Reply:
The stamping context currently exposes the path to input values based on their type. When the input type becomes a degree of freedom, we no longer control the top-level key space, collisions between types and other existing top-level keys can happen.

For example, these values are currently in our samples

https://github.com/vmware-tanzu/cartographer/blob/6d592c8f8d7ad2a961f556976f9bd33460f9b726/examples/shared/app-operator/app-deploy-template.yaml#L80

https://github.com/vmware-tanzu/cartographer/blob/6d592c8f8d7ad2a961f556976f9bd33460f9b726/examples/shared/app-operator/kpack-image-template.yaml#L47

Reply:
^ we also have some aliasing in case there's only one of those (e.g., if you have only a single source, then rather than `$(sources.source.url)$`, one can do `$(source.url)$`